man/shatag.1
.TH shatag 1 26.08.2010 "Shatag 0.1" "Shatag"

.SH NAME

shatag - tag files with their SHA-256 checksums

.SH SYNOPSIS
.B shatag
.nh
[\fB\-fhlLqrstuv0\fR]
[\fB\-d \fIDATABASE\fR]
[\fB\-n \fINAME\fR]
[\fB\-R \fINAME\fR]...
[\fIFILES\fR]...
.hy

.SH DESCRIPTION
.B shatag
is a tool for computing and caching SHA-256 file checksums, and efficiently search
for identical file across systems. Checksums are stored using the POSIX Extended
Attributes filesystem facility, and are preserved when files are moved or renamed.
Checksums can be fetched from a remote host and stored in an sqlite database
for fast lookups.

.SH OPTIONS
When invoked with no options,
.B shatag
just displays the cached, valid checksums. If no files are specified, it applies to all non-hidden files
in the current directory. The output format is identical to the one of the
.B sha256sum
command.

.IP "\fB\-0, \-\-null\fP"
Instead of outputting one record per line (like
.B sha256sum
does,) separate records with null characters.

.IP "\fB\-c, \-\-canonical\fP"
Show canonical (full path) file names.

.IP "\fB\-d \fIDATABASE\fR, \fB\-\-database \fIDATABASE\fP"
Set the path of the SQLite database to query when using 
.B "\-l" 
, 
.B "\-L"
or
.B "\-p"
(The default path is $HOME/.shatagdb, overridable from the config file)

Instead of a file name, a PostgreSQL database can be specified with a prefix of "pg:"
followed by a psycopg2 DSN string, like:

"pg:dbname=shatag user=myuser password=mypassword host=192.168.1.3"

.IP "\fB\-f, \-\-force\fP"
When running with
.B -t
or
.B -u
, force recompute the checksum and overwrite the old one, even if the timestamp indicates
a good checksum.

.IP "\fB\-h, \-\-help\fP"
Displays the help message

.IP "\fB\-l, \-\-lookup\fP"
Instead of displaying the checksums, look them up against the local database
and indicate if the file exists. A
.B "yellow -"
mark indicates that the file does not exist somewhere else, a 
.B "green ="
that the file exists at one or several remote locations, a
.B "red +"
that the file has a duplicate on the local system, and a
.B "magenta *"
that the file is empty.

.IP "\fB\-L, \-\-lookup\-verbose\fP"
Instead of displaying the checksums, look them up against the local database.
Print all the known remote locations for identical files.

.IP "\fB\-n \fINAME\fR, \fB\-\-name \fINAME\fP"
Name of local storage (defaults to canonical local host name). This needs to be correct if
the local database contains entries for this own host.

.IP "\fB\-p, \-\-put\fP"
Record found tags in the database, for duplicate detection.

.IP "\fB\-q, \-\-quiet\fP"
Do not display the valid checksums when they are found.

.IP "\fB\-r, \-\-recursive\fP"
Recurse trough subdirectories

.IP "\fB\-R \fINAME\fR, \fB\-\-remote \fINAME\fP"
When using
.B \-l
or
.B \-L
, This is used to restrict the set of remote names to consider. If present, other storages will be ignored.

.IP "\fB\-s, \-\-scrub\fP"
Recompute the checksum even if the timestamp indicates it would not be needed, and report inconsistencies.
Useful to detect silent corruption.

.IP "\fB\-t, \-\-tag\fP"
Compute new checksums for files that don't have one, or when it is outdated.

.IP "\fB\-u, \-\-update\fP"
Recompute the outdated checksums only. Be aware that this can behave counter-intuitively;
outdated checksums will only exists for files that have been appended to or partially modified.
Many programs dealing with small files (some well-known text editors, notably) will overwrite
the whole file when saving, and the new file will be lacking a checksum entirely. For these
cases, use
.B "\-t" instead.

.IP "\fB\-v, \-\-verbose\fP"
Report encountered files that have an outdated or missing checksum.

.SH EXAMPLES
.LP
Retag a whole directory and record everything to the database:
.RS
.nf
\fBshatag \-pqrt .\fP
.fi
.RE
.LP
Check files in the current directory for remote duplicates:
.RS
.nf
\fBshatag \-l\fP
.fi
.RE
.LP
Show alternate locations for duplicates of a single file:
.RS
.nf
\fBshatag \-L somefile\fP
.fi
.RE

.SH FILES

.TP
~/.shatagrc
YAML configuration file. Currently has only two possible configuration keys: "database", which sets the database path (by default, ~/.shatagdb) and "name" for the volume name in the database (default to canonical host name.)

Examples:

database: /var/lib/shatag.db   # sqlite3 backend
database: "pg: dbname=shatag host=localhost user=shatag password=xxxsecretpasswordxxx"  # postgres backend
database: http://service.com/shatag   # http backend
database: insecure-https://service.com/shatag  # http backend, skip ssl certificate verification

.SH BUGS
Support for non-ASCII filenames across systems of different and/or inconsistent encodings have not been fully tested.

Not all option combinations are sensible.

.SH "REPORTING BUGS"
Report shatag bugs to the bugtracker at http://bitbucket.org/maugier/shatag,

.SH "SEE ALSO"
shatag-add(1), shatagd(1)


shatag/cli/shatag.py
#!/usr/bin/env python3
# Copyright 2010-2016 Maxime Augier
# Distributed under the terms of the GNU General Public License

import argparse
import os
import os.path
import shatag
import sys

config = shatag.Config()

parser = argparse.ArgumentParser(description='Display and update xattr-based checksums.')
parser.add_argument('-c', '--canonical', action='store_true', help='Output canonical file names.')
missing_group = parser.add_mutually_exclusive_group()
missing_group.add_argument('-t', '--tag', action='store_true', help='add checksums to untagged files')
missing_group.add_argument('-u', '--update', action='store_true', help='update outdated checksum')
good_group = parser.add_mutually_exclusive_group()
good_group.add_argument('-f', '--force', action='store_true', help='force recompute all checksums, overwrite inconsistencies')
good_group.add_argument('-s', '--scrub', action='store_true', help='force recompute all checksums, report inconsistencies')
parser.add_argument('-q', '--quiet', action='store_true', help='do not output checksums')
parser.add_argument('-v', '--verbose', action='store_true', help='report missing/outdated checksums')
parser.add_argument('-r', '--recursive', action='store_true', help='inspect directories recursively')
parser.add_argument('-0', '--null', action='store_true', help='separate output records with NULLs instead of newlines')
lookup_group = parser.add_mutually_exclusive_group()
lookup_group.add_argument('-l', '--lookup', action='store_true', help='look up files in the database, color/symbol-code them')
lookup_group.add_argument('-L', '--lookup-verbose', action='store_true', help='verbosely list duplicate locations')
parser.add_argument('-p', '--put', action='store_true', help='add tags to database')
parser.add_argument('-d', '--database', metavar='DB', help='database path for -l/-L mode', default=config.database)
parser.add_argument('-b', '--backend', metavar='BACKEND', help='backend for local tag storage', default=config.backend)
parser.add_argument('-n', '--name', metavar='NAME', help='name of local storage for -l/-L mode', default=config.name)
parser.add_argument('-R', '--remote', metavar='NAME', action='append', help='remote storage to consider for -r/-R mode')
parser.add_argument('files', metavar='FILE', nargs='*', help='files to checksum')


def main():

    args = parser.parse_args()

    if args.files == []:
        if args.tag or args.update:
            print('shatag: Error: file name(s) required when using -t or -u', file=sys.stderr)
        else:
            args.files = filter(lambda x: (args.recursive or os.path.isfile(x))
                                and not x.startswith('.'), os.listdir('.'))

    if (args.quiet and (not args.verbose) and not (args.update or args.tag or args.force or args.scrub or args.put)):
        print('shatag: Warning: this combination of flags does not do anything.', file=sys.stderr)
        exit(1)

    if (args.quiet and args.null):
        print('shatag: Warning: --null useless when using --quiet', file=sys.stderr)

    end = args.null and '\0' or '\n'
    store = None

    if args.lookup or args.lookup_verbose or args.put:
        store = shatag.Store(name=args.name, url=args.database)

    backend = shatag.backend(args.backend)

    def process(filename):
        try:

            if args.recursive:
                if os.path.isdir(filename) and not os.path.islink(filename):
                    [process(os.path.join(filename, e)) for e in os.listdir(filename) if not e.startswith('.')]
                    return

            else:
                if not os.path.isfile(filename):
                    print('shatag: Warning: {0} is not a file'.format(filename), file=sys.stderr)
                    return

            file = backend.file(filename)
            if args.verbose:
                file.verbose()

            if file.state == 'good' and (args.force or args.scrub):
                file.rehash(args.canonical, args.scrub)

            if args.update:
                file.update(canonical=args.canonical)
            elif args.tag:
                file.tag(canonical=args.canonical)

            try:
                if not (args.quiet or args.lookup or args.lookup_verbose):
                    file.fsprint(file.show(canonical=args.canonical), end=end)

                if args.put:
                    store.put(file)

                if args.lookup:
                    file.fsprint(store.lookup(file, remotenames=args.remote).pretty())

                if args.lookup_verbose:
                    r = store.lookup(file)
                    print("{0}:".format(file.path(canonical=args.canonical)))
                    for (name, path) in (r.local + r.remote):
                        print("\t{0}:{1}".format(name, path))

            except shatag.NoChecksum:
                pass

        except IOError as e:
            print('shatag: "{0}": IOError {1}: {2}'.format(filename.encode(), e.errno, e.strerror), file=sys.stderr)
            if e.errno == 32:
                raise
            if e.errno == 95:
                print('shatag: "{0}": Operation Unsupported. Did you forget to enable user_xattr ?'.format(filename.encode()), file=sys.stderr)
        except OSError as e:
            print('shatag: {0}'.format(e), file=sys.stderr)

    for filename in args.files:
        if args.put and args.recursive:
            store.clear(os.path.abspath(filename))
        process(filename)

    if args.put:
        store.commit()
